runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/data/GrpcDataService.java

```diff
@@ -79,8 +81,12 @@ private GrpcDataService(ExecutorService executor) {
   public StreamObserver<BeamFnApi.Elements> data(
       final StreamObserver<BeamFnApi.Elements> outboundElementObserver) {
     LOG.info("Beam Fn Data client connected.");
+    // TODO: Consider whether here, too, we need fancier synchronization like that done
```

Comment:
Pass in the StreamObserverFactory now.

Reply:
Done, added ServerStreamObserverFactory (similar to the one in Dataflow worker).